### ISSUE TITLE
 Tests no longer pass on Rakudo HEAD

### ISSUE THREAD

**Issue by AlexDaniel** [raku-community-modules/OO-Actors#6]
Bisectable points at (2017-09-27) https://github.com/rakudo/rakudo/commit/9af5607d523ec9ba40921ad3b09206cf74f3ce37

Output before and after:
```
c: 9af5607d523^,9af5607d523 https://gist.githubusercontent.com/AlexDaniel/1799a7700a416b216c32b8347dc1caaf/raw/04661ce642abc541ad80bbf38aa4a98abd59f5ad/basic.p6
```

```
¦9af5607d523^:
«1..3
ok 1 - Actor does Actor role
ok 2 - Made many calls to Actor over many threads
ok 3 - Correct number of entries»

¦9af5607:
«1..3
ok 1 - Actor does Actor role
«timed out after 10 seconds» «exit signal = SIGHUP (1)»»
```

See also: https://github.com/jnthn/p6-test-scheduler/issues/3

**Comment by AlexDaniel** [comment]
Ok, this is a really weird one, but the problem is in the module itself (or so I believe).

So it does $*SCHEDULER.cue(stuff) hundreds of times
https://github.com/jnthn/oo-actors/blob/886b8ace0b6c18f8790b8f3dc7c96c481453e311/lib/OO/Actors.pm#L15

where each “stuff” has an infinite loop:
https://github.com/jnthn/oo-actors/blob/886b8ace0b6c18f8790b8f3dc7c96c481453e311/lib/OO/Actors.pm#L22-L43

and that infinite loop never ends, never sleeps, and never reasonably blocks on anything. It seems that this starves the main thread and keeps it from proceeding. If you use lower number here (e.g. 15 instead of 100):
https://github.com/jnthn/oo-actors/blob/886b8ace0b6c18f8790b8f3dc7c96c481453e311/t/basic.t#L25

Then tests pass, but of course things are really slow so you might need to wait a few seconds…

In any case, does not look like a rakudo issue at all. That said, it makes me wonder how it managed to work in the past.

**Comment by AlexDaniel** [comment]
Apparently this is a flapper because it managed to pass in toaster on 2018.01 (once).

**Comment by tbrowder** [comment]
It’s not passing on my Toaster test, it just hangs and stalls the whole process. It cannot be killed by Proc::Q as used by module Toaster.

It also hangs when testing locally with Rakudo 2018.03.

**Comment by marcoonroad** [comment]
Hooray! GitHub is awesome! I will make a PR once the tests pass on Travis-CI [here](https://travis-ci.org/marcoonroad/oo-actors).

Let's pray! :pray:

**Comment by marcoonroad** [comment]
I'm using Channels & Threads as a drop-in replacement. I must also test the garbage collection to see whether the things are properly destroyed. :question: :question: :question: :question:

(Anyways, the tests - not the build per se - run quite fast.)

**Comment by jnthn** [comment]
Fixed in d0445f4446e245375d2e using `Lock::Async`, which has been in Rakudo for quite a while by now and solves the tricky problem that the broken code in this module was trying to solve, except it does it correctly (and if for some reason it isn't correct, we'll find out about it, since it's the mechanism that powers all `Supply` concurrency!)
